rfcs/003-strings.md

```diff
@@ -0,0 +1,325 @@
+
+The main conversions afforded to these types are to:
+
+* `&[u16]` or `&[u8]` (depending if the given Windows type is a wide string or not)
+* `String`
+
+Conversion of most types to `CString` and `OString` doesn't make much sense as all Windows types are already FFI safe and those Rust types are almost never needed outside of FFI. However, the user should be given access to the underlying buffers (including the trailing nul byte), and there should be conveniences to converting to `String` since this unlocks the entire rich `std` library functionality for string manipulation.
+
+All string types that aren't `HSTRING` can only expose unsafe functions as it is not well known that they point to valid memory. Additionally, each type should allow conversion to a raw pointer to avoid having to use `std::mem::transmute` in cases where an API expects a raw pointer.
+
+#### From Rust to Windows types
+
+It is common for Rust functionality to want to interact with some Windows APIs which require the Windows string types.
+
+Converting from Rust types usually requires copy the UTF-8 encoded bytes to UTF-16. This can be done with any type that can be referenced as a `str`, so it makes sense to provide `From<&T> where T: AsRef<str>` for `HSTRING` as `HSTRING` will ensure the newly allocated buffer is freed when appropriate.
+
+All the other Windows string types do not have clear ownership semantics and thus providing convenient conversions might risk leading the user to leak memory. This is the same reason logically borrowed types like `CStr` don't allow converting from owned types like `String`. As such, only direct conversions will be provided: conversion from `*const u8` and `*const u16`.
+
+#### API Proposal
+
+With the above in mind, here is what the API should look like:
+
+```rust
+impl HSTRING {
+    // String data without the trailing 0
+    fn as_wide(&self) -> &[u16] {}
+    // String data with the trailing 0
+    fn as_wide_with_nul(&self) -> &[u16] {}
+    fn to_string(&self) -> Result<String, FromUtf16Error> {}
+    fn to_string_lossy(&self) -> String {}
+    fn as_ptr(&self) -> *const c_void;
+}
+
+// Display shows the string with non-valid utf16 replaced with �
+impl Display for HSTRING {}
+// Same as Display but surrounded by ""
+impl Debug for HSTRING {}
+
+// Uses the `to_string_lossy` function
+impl From<HSTRING> for String {}
+// Uses the `to_string` function
+impl TryFrom<HSTRING> for STRING {
+    type Error = FromUtf16Error;
+}
+impl<T: ?Sized + AsRef<str>> From<&'_ T> for HSTRING {}
+impl<T: ?Sized + AsRef<str>> TryFrom<&'_ T> for HSTRING {
+    type Error = FromUtf16Error;
+}
+
+// --------------------------------
+
+/// The wide equivalent to std::ffi::CString;
+#[repr(transparent)]
+struct CWString(..);
```

Comment:
Should there be a matching borrowed `CWStr`, like `CString` and `CStr`?

Reply:
@rylev 
One reason would be so there isn't a double-indirection.
It can also be used to convert a `PCWSTR` into a safely-used reference without having to copy or own the data. (See for example `CStr::from_ptr`)